### PR TITLE
WIP: sweepers inside conf/model.yaml and ddm_trainer.py

### PR DESCRIPTION
**Pull request by akzaidi**
# Pull Request Template

## PR Song

[Nujabes - sea of cloud](https://www.youtube.com/watch?v=dBYDcZMCRnY&ab_channel=PB)

## Progress Items

I've only added and done a quick test for sweepers with the xgboost models and the default conf template. I'll test the remainder functions and add tests + documentation.

## Description

This PR includes work to include the sweeper functionality directly into `ddm_trainer` and into the `conf` files.

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## How Has This Been Tested?

Currentlly tested on xgboost models with the default configuration. See the updated syntax for `sweep` in `conf/model/xgboost.yaml`. Further tests to be added shortly.

## Reviewer Tasks

List here all the tasks the reviewer needs to review:

- [x] Test `examples/`
- [x] Test with more hyperparameters

**Test Configuration**:
* OS: macOS
* Python Version: 3.7

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by jillianmclements** [comment]
Should we set the logger to not log UserWarnings? I see this one in the output over 10 times:
`/opt/anaconda3/envs/ddm/lib/python3.7/site-packages/xgboost/data.py:106: UserWarning: Use subset (sliced data) of np.ndarray is not recommended because it will generate extra copies and increase memory consumption`

**Comment by jillianmclements** [comment]
The ability to add `verbose=2` to TuneSearchCV might be interesting to some.

**Comment by jillianmclements** [comment]
The default CV in `TuneSearchCV` in default 5-fold cross validation. This might not be ideal for sequential data. Without the sweep enabled, is there a default CV in `fit`?

**Comment by akzaidi** [comment]
> The ability to add `verbose=2` to TuneSearchCV might be interesting to some.

ooh lemme see if I can add a verbosity option.

**Comment by akzaidi** [comment]
> Should we set the logger to not log UserWarnings? I see this one in the output over 10 times:
> `/opt/anaconda3/envs/ddm/lib/python3.7/site-packages/xgboost/data.py:106: UserWarning: Use subset (sliced data) of np.ndarray is not recommended because it will generate extra copies and increase memory consumption`

hmm it appears these aren't directly called by `logging`, so I'll have to try and intercept them. Lemme see what I can do.

**Comment by akzaidi** [comment]
> The default CV in `TuneSearchCV` is default 5-fold cross validation. This might not be ideal for sequential data. Can we add the ability specify a test set?
> 
> Out of curiosity, is there a default CV in `fit` without the sweep enabled?

Ah yeah, all of our training and sweeping loops assume the data is already partitioned into [x,y] sets and we ignore any temporal dependence. Let me work on adding the logic for fixed train/test splits or allow the user to split directly by a temporal column like `episode` so they could then do something like block-sampling.

The `fit` method has no sweep built-in at the moment.

**Comment by jillianmclements** [comment]
> > The default CV in `TuneSearchCV` is default 5-fold cross validation. This might not be ideal for sequential data. Can we add the ability specify a test set?
> > Out of curiosity, is there a default CV in `fit` without the sweep enabled?
> 
> Ah yeah, all of our training and sweeping loops assume the data is already partitioned into [x,y] sets and we ignore any temporal dependence. Let me work on adding the logic for fixed train/test splits or allow the user to split directly by a temporal column like `episode` so they could then do something like block-sampling.
> 
> The `fit` method has no CV built-in at the moment.

That's fair and I think you're right, maybe it's not a big deal. I like your idea for fixed splits though (not just for sweep, but for reporting the results from `fit` in the terminal). Thanks!

**Comment by akzaidi** [comment]
@jillianmclements sorry this took so long, but I think it's ready enough!

File: base.py [hunk 1]
```diff
@@ -530,20 +544,135 @@ def group_per_episode(self, X, y = None, episode_ids = None):
...
+        with mlflow.start_run() as run:
+            search.fit(X, y)
+        self.model = search.best_estimator_
+        results_df = pd.DataFrame(search.cv_results_)
+        if not pathlib.Path(results_csv_path).parent.exists():
+            pathlib.Path(results_csv_path).parent.mkdir(exist_ok=True, parents=True)
+        logger.info(f"Saving sweeping results to {results_csv_path}")
+        results_df.to_csv(results_csv_path)
+        logger.info(f"Best hyperparams: {search.best_params_}")
+
```
Comment (jillianmclements) [inline base.py:646]: Request to output final score as well: 
`logger.info(f"Best score: {search.best_score_}")`
